dev/io.openliberty.microprofile.reactive.messaging.internal/src/io/openliberty/microprofile/reactive/messaging/internal/RMAsyncProviderFactoryImpl.java

```diff
@@ -0,0 +1,226 @@
+                                                                                  Collections.singletonMap(WSContextService.THREAD_CONTEXT_PROVIDER,
+                                                                                                           "com.ibm.ws.classloader.context.provider"),
+                                                                                  Collections.singletonMap(WSContextService.THREAD_CONTEXT_PROVIDER,
+                                                                                                           "com.ibm.ws.javaee.metadata.context.provider"),
+                                                                                  Collections.singletonMap(WSContextService.THREAD_CONTEXT_PROVIDER,
+                                                                                                           "com.ibm.ws.security.context.provider"),
+    };
+
+    /**
+     * The built-in context service
+     */
+    @Reference(target = "(service.pid=com.ibm.ws.context.manager)", policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile WSContextService builtInContextService;
+
+    /**
+     * The default context service provided by the concurrent feature
+     */
+    @Reference(target = "(id=DefaultContextService)", cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile WSContextService defaultContextService;
+
+    /**
+     * A map of named context services configured in the server.xml
+     * <p>
+     * Maps from the context service ID to the context service.
+     * <p>
+     * Updated dynamically if context services are added or removed while the server is running.
+     */
+    private final Map<String, WSContextService> namedContextServices = new ConcurrentHashMap<>();
+
+    /**
+     * The liberty global thread pool executor
+     */
+    @Reference(target = "(component.name=com.ibm.ws.threading)", policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile ExecutorService globalExecutor;
+
+    /**
+     * A scheduled executor which delegates to the global thread pool
+     */
+    @Reference(target = "(deferrable=false)", policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    private volatile ScheduledExecutorService scheduledExecutor;
+
+    /**
+     * Updates {@link #namedContextServices} when a new service is added
+     */
+    @Reference(service = WSContextService.class, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.MULTIPLE)
+    protected void addContextService(WSContextService contextService, Map<String, Object> properties) {
+        String name = getContextServiceName(properties);
+        if (name != null) {
+            namedContextServices.put(name, contextService);
+        }
+    }
```

Comment:
My thinking was:
- we need to find a context service by name regularly (usually at least once per message processed)
  - we need to look it up every time we use it because context services defined in the server.xml can change dynamically
- the total number of context services will be small
- context services don't come and go very often

I assumed (without testing) that it would be quicker to maintain a map from name to context service than to do an OSGi service lookup every time we wanted to use it. I wasn't aware of an easy way to do a lookup the first time we need a context service with a given name and then track it to see if it goes away.

I guess we could create a `ServiceTracker` every time we need a service with a given name and maintain a map of name to corresponding `ServiceTracker`? The tracker could remove itself from the map if the service goes away. I was assuming that the total number of context services would be small so the overhead of maintaining the map would be insignificant and didn't consider doing something more complex.

The suggestion to use a filter to ignore services without an ID is a reasonable one.

Reply:
I made the comment because it seemed odd to me to be collecting and tracking extra services that aren't needed, but your assumptions here are valid and so your approach does make sense.  Most other users of ContextService have server.xml config (or app config) that directly references a single instance or defaults, in which case the dependency can be enforced in OSGi on the one instance that is needed.  You have a more complex situation with the ability to receive requests for multiple different context services.